### PR TITLE
dmc manipulator

### PR DESCRIPTION
**Pull request by yufansong**
#103

### CODE REVIEW

File: envpool/mujoco/dmc/manipulator.h [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+    std::array<mjtNum, 3> diff = {data_->site_xpos[id_site_1 * 3 + 0] -
+                                      data_->site_xpos[id_site_2 * 3 + 0],
+                                  data_->site_xpos[id_site_1 * 3 + 1] -
+                                      data_->site_xpos[id_site_2 * 3 + 1],
+                                  data_->site_xpos[id_site_1 * 3 + 2] -
+                                      data_->site_xpos[id_site_2 * 3 + 2]};
+    return std::sqrt(diff[0] * diff[0] + diff[1] * diff[1] + diff[2] * diff[2]);
+  }
+
+  [[nodiscard]] float IsClose(double distance) const {
```
Comment (yufansong) [inline envpool/mujoco/dmc/manipulator.h:249]: Do not why `make lint` require adding `[[nodiscard]]`

File: envpool/mujoco/dmc/manipulator.h [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+
+  std::array<mjtNum, 4> Body2dPose(const std::string& body_names) {
+    int id = mj_name2id(model_, mjOBJ_XBODY, body_names.c_str());
+    // self.named.data.xpos[body_names, ['x', 'z']]
+    // self.named.data.xquat[body_names, ['qw', 'qy']]
+    return {data_->xpos[id * 3 + 0], data_->xpos[id * 3 + 2],
+            data_->xquat[id * 4 + 0], data_->xquat[id * 4 + 2]};
+  }
+
+  std::string MakeManipulatorModel(const Spec& spec) {
```
Comment (Trinkle23897) [inline envpool/mujoco/dmc/manipulator.h:318]: you can move this part to GetManipulatorXML(spec.config["base_path"_], spec.config["task_name"_]);

File: envpool/mujoco/dmc/manipulator.h [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        content = ReplaceRegex(content, k_all_props);
+      }
+    }
+    return content;
+  }
+  static std::string ReplaceRegex(const std::string& content,
+                                  const std::string& unused_prop) {
+    std::ostringstream pattern_ss;
+    pattern_ss << "<body name=\"" << unused_prop
+               << R"("((?!</body>)[\s\S])+</body>)";
```
[the remark this answers is not in the document]
Comment (Trinkle23897) [reply envpool/mujoco/dmc/manipulator.h:369]: https://github.com/zeux/pugixml

**Comment by Trinkle23897** [comment]
```c++
#include <bits/stdc++.h>

#include "pugixml.hpp"

struct xml_string_writer: pugi::xml_writer
{
    std::string result;

    virtual void write(const void* data, size_t size)
    {
        result.append(static_cast<const char*>(data), size);
    }
};
std::string node_to_string(pugi::xml_node node)
{
    xml_string_writer writer;
    node.print(writer);

    return writer.result;
}

std::string GetManipulatorXML(const std::string &base_path,
                              const std::string &task_name)
{
    std::vector<std::string> unused_props;
    if (task_name == "bring_ball") {
        unused_props = {"slot", "target_peg", "cup", "peg"};
    } else if (task_name == "bring_peg") {
        unused_props = {"slot", "target_ball", "cup", "ball"};
    } else if (task_name == "insert_ball") {
        unused_props = {"slot", "target_peg", "peg"};
    } else if (task_name == "insert_peg") {
        unused_props = {"target_ball", "cup", "ball"};
    }
    std::string filename = base_path + "/manipulator.xml";
    std::ifstream ifs(filename);
    pugi::xml_document doc;
    doc.load(ifs);
    for (auto prop : unused_props) {
        std::string xpath = "//body[@name='" + prop + "']";
        pugi::xml_node node = doc.select_node(xpath.c_str()).node();
        auto parent = node.parent();
        parent.remove_child(node);
    }
    return node_to_string(doc);
}

int main(int argc, char **argv) {
    std::cout << GetManipulatorXML("dmc", argv[1]) << std::endl;
}
```

**Review by Trinkle23897 — APPROVED** [review]

**Pull request merged**